File: src/daytrade/quotes.py
"""9:00 の気配・現在値の取得元。

日本株のリアルタイム気配を返す取得元は今のところ無い。取得元は差し替え可能に
してあるので、証券会社の市場データ API を使うならここに足す。
``daytrade quotes 7203 9984`` で疎通を確かめる。

- ``yfinance``: Yahoo Finance。東証は **20 分遅れ**なので寄付の判断には使えない。
  ``delayed=True`` を付けて返し、``open`` は既定で拒否する（検証・dry-run 用）
- ``csv``: ``symbol,price[,at]`` のファイル。手で入れた気配や、別経路で取った値を流す
"""

from __future__ import annotations

import csv
import datetime as dt
import logging
from collections.abc import Iterable
from decimal import Decimal, InvalidOperation
from pathlib import Path
from typing import Any, ClassVar, Protocol

from daytrade.select import Quote
from wbcore.clock import ensure_utc, now_utc
from wbcore.credentials import Environment
from wbcore.logging import get_logger

log = get_logger(__name__)


class QuoteError(RuntimeError):
    """気配を取れなかった。"""


class QuoteSource(Protocol):
    @property
    def name(self) -> str: ...

    def fetch(self, symbols: Iterable[str]) -> dict[str, Quote]:
        """銘柄 → 気配。取れなかった銘柄は含めない。"""


def _decimal(value: Any) -> Decimal | None:
    if value in (None, "", "null"):
        return None
    try:
        result = Decimal(str(value))
    except InvalidOperation:
        return None
    return result if result > 0 else None


class YFinanceQuotes:
    """Yahoo Finance（20 分遅れ）。``delayed=True`` を付けて返す。検証・dry-run 用。"""

    name: ClassVar[str] = "yfinance"

    def fetch(self, symbols: Iterable[str]) -> dict[str, Quote]:
        import yfinance as yf

        wanted = list(dict.fromkeys(s for s in symbols if s))
        if not wanted:
            return {}
        tickers = [f"{s}.T" for s in wanted]
        # yfinance は取れない銘柄ごとに error を吐く（上場廃止など）。こちらで missing として
        # 数えるので、code の無い error でログを埋めないよう黙らせる
        logging.getLogger("yfinance").setLevel(logging.CRITICAL)
        try:
            frame = yf.download(
                tickers, period="1d", interval="1m", group_by="ticker", progress=False, threads=True
            )
        except Exception as exc:
            raise QuoteError(f"yfinance からの取得に失敗: {exc}") from exc
        found: dict[str, Quote] = {}
        for symbol, ticker in zip(wanted, tickers, strict=True):
            try:
                series = frame[ticker]["Close"] if len(tickers) > 1 else frame["Close"]
            except KeyError, TypeError:
                continue
            series = series.dropna()
            if series.empty:
                continue
            price = _decimal(series.iloc[-1])
            if price is None:
                continue
            stamp = series.index[-1].to_pydatetime()
            at = ensure_utc(stamp)
            found[symbol] = Quote(symbol=symbol, price=price, at=at, source=self.name, delayed=True)
        return found


class CsvQuotes:
    """``symbol,price[,at]`` の CSV。``at`` は ISO 8601（無ければ今）。"""

    name: ClassVar[str] = "csv"

    def __init__(self, path: Path) -> None:
        self.path = Path(path)

    def fetch(self, symbols: Iterable[str]) -> dict[str, Quote]:
        wanted = set(symbols)
        if not self.path.is_file():
            raise QuoteError(f"気配の CSV がありません: {self.path}")
        found: dict[str, Quote] = {}
        with self.path.open(encoding="utf-8", newline="") as handle:
            for row in csv.DictReader(handle):
                symbol = (row.get("symbol") or "").strip()
                price = _decimal(row.get("price"))
                if symbol not in wanted or price is None:
                    continue
                raw_at = (row.get("at") or "").strip()
                at = ensure_utc(dt.datetime.fromisoformat(raw_at)) if raw_at else now_utc()
                found[symbol] = Quote(symbol=symbol, price=price, at=at, source=self.name)
        return found


class TachibanaQuotes:
    """立花証券 e支店 API の時価問合（``CLMMfdsGetMarketPrice``、1 リクエスト最大 120 銘柄）。

    寄付後は始値（``pDOP``）、無ければ現在値（``pDPP``）。ブローカーの接続（ログイン・
    仮想URL）をそのまま使う。
    """

    name: ClassVar[str] = "tachibana"

    def __init__(self, broker: Any) -> None:
        self._broker = broker

    @classmethod
    def connect(cls, env: Environment) -> TachibanaQuotes:
        from wbcore.broker.tachibana import TachibanaBroker
        from wbcore.domain.models import Market

        return cls(TachibanaBroker.connect(env, market=Market.JP))

    def fetch(self, symbols: Iterable[str]) -> dict[str, Quote]:
        from wbcore.broker.base import BrokerError

        wanted = list(dict.fromkeys(s for s in symbols if s))
        try:
            rows = self._broker.market_prices(wanted)
        except BrokerError as exc:
            raise QuoteError(f"立花証券の時価取得に失敗: {exc}") from exc
        found: dict[str, Quote] = {}
        for symbol, row in rows.items():
            price = _decimal(row.get("open")) or _decimal(row.get("last"))
            if price is None:
                continue
            found[symbol] = Quote(symbol=symbol, price=price, at=row["at"], source=self.name)
        return found


def quote_source(name: str, env: Environment, *, quote_file: Path | None = None) -> QuoteSource:
    """設定の名前から取得元を組み立てる。"""
    if name == "tachibana":
        return TachibanaQuotes.connect(env)
    if name == "yfinance":
        return YFinanceQuotes()
    if name == "csv":
        if quote_file is None:
            raise ValueError('quote_source = "csv" には quote_file が必要です')
        return CsvQuotes(quote_file)
    raise ValueError(f"未知の quote_source: {name!r}（tachibana / yfinance / csv）")

File: cmd/daytrade/quotes.go
package main

import (
	"encoding/json"
	"fmt"

	dtquotes "github.com/lovemoneyhotspring/jstock-go/pkg/daytrade/quotes"
	"github.com/lovemoneyhotspring/jstock-go/pkg/wbcore/clock"
	"github.com/spf13/cobra"
)

func newQuotesCmd() *cobra.Command {
	var sourceFlag, quoteFileFlag, columnsFlag string
	var rawFlag bool
	cmd := &cobra.Command{
		Use:   "quotes SYMBOL...",
		Short: "気配の取得元の疎通を確かめる（寄付の判断に使える鮮度かはここで見る）",
		Args:  cobra.MinimumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if rawFlag {
				return showRawQuotes(args, columnsFlag)
			}
			cfg, err := loadConfig()
			if err != nil {
				return err
			}
			found, err := fetchQuotes(cfg, args, sourceFlag, quoteFileFlag)
			if err != nil {
				return err
			}
			now := clock.NowUTC()
			zone := clock.MustZone(appSettings.Timezone)
			for _, symbol := range args {
				q, ok := found[symbol]
				if !ok {
					fmt.Printf("  %s: 取れませんでした\n", symbol)
					continue
				}
				flag := ""
				if q.Delayed {
					flag = "（遅延）"
				}
				fmt.Printf("  %s: %s 円  %s  %d 秒前 %s%s\n",
					symbol, yen(q.Price), clock.Fmt(q.At, zone, true),
					int(now.Sub(q.At).Seconds()), q.Source, flag)
			}
			return nil
		},
	}
	cmd.Flags().StringVar(&sourceFlag, "source", "", "tachibana / csv")
	cmd.Flags().StringVar(&quoteFileFlag, "quote-file", "", "csv のときのファイル")
	cmd.Flags().BoolVar(&rawFlag, "raw", false, "時価問合の応答をそのまま JSON で出す（立花のみ。板の列名の確認用）")
	cmd.Flags().StringVar(&columnsFlag, "columns", "",
		"--raw で取りに行く列（sTargetColumn。既定は始値・現在値・現在値時刻・前日終値）")
	return cmd
}

// showRawQuotes は時価問合の応答を解釈せずに出す。
//
// 板の列（気配値・気配数量・特別気配）を記録に足す前に、「その列名で何が返るか」を
// 実機で確かめるための口（docs/OPENING_DATA.md「実機で確かめること」）。仕様書に
// 無い列を指定しても、空で返るのか誤りになるのかがそのまま見える。
func showRawQuotes(symbols []string, columns string) error {
	source, err := dtquotes.Connect(appSettings.Env, appSettings.DotenvMap, appSettings.StateDir)
	if err != nil {
		return err
	}
	rows, err := source.Broker.MarketPricesRaw(symbols, columns)
	if err != nil {
		return err
	}
	if len(rows) == 0 {
		fmt.Println("  応答に行がありません（銘柄コードか列の指定を確かめてください）")
		return nil
	}
	out, err := json.MarshalIndent(rows, "", "  ")
	if err != nil {
		return err
	}
	fmt.Println(string(out))
	return nil
}
